xds/internal/balancer/edsbalancer/balancergroup.go

```diff
@@ -167,50 +287,41 @@ func (bg *balancerGroup) changeWeight(id internal.Locality, newWeight uint32) {
 		return
 	}
 	pState.weight = newWeight
-	// Update state and picker to reflect the changes.
-	bg.cc.UpdateBalancerState(buildPickerAndState(bg.idToPickerState))
+	if bg.pickerStarted {
+		// Update state and picker to reflect the changes.
+		bg.cc.UpdateBalancerState(buildPickerAndState(bg.idToPickerState))
+	}
 }
 
 // Following are actions from the parent grpc.ClientConn, forward to sub-balancers.
 
 // SubConn state change: find the corresponding balancer and then forward.
 func (bg *balancerGroup) handleSubConnStateChange(sc balancer.SubConn, state connectivity.State) {
 	grpclog.Infof("balancer group: handle subconn state change: %p, %v", sc, state)
-	bg.mu.Lock()
-	var b balancer.Balancer
+	bg.incomingMu.Lock()
 	if id, ok := bg.scToID[sc]; ok {
 		if state == connectivity.Shutdown {
 			// Only delete sc from the map when state changed to Shutdown.
 			delete(bg.scToID, sc)
 		}
-		b = bg.idToBalancer[id]
-	}
-	bg.mu.Unlock()
-	if b == nil {
-		grpclog.Infof("balancer group: balancer not found for sc state change")
-		return
-	}
-	if ub, ok := b.(balancer.V2Balancer); ok {
-		ub.UpdateSubConnState(sc, balancer.SubConnState{ConnectivityState: state})
-	} else {
-		b.HandleSubConnStateChange(sc, state)
+		bg.outgoingMu.Lock()
```

Comment:
This mutex might cause a deadlock if balancer's `handleSubConnStateChange` calls back to balancer group to create a SubConn... I moved it out from the mutex.